File: java/ods/XFastTrie.java
package ods;

public class XFastTrie<Node extends XFastTrie.Nöde<Node,T>, T> 
	extends BinaryTrie<Node, T> {

	/**
	 * The hash tables used to store prefixes
	 */
	USet<Node>[] t;
	
	protected static class Nöde<N extends Nöde<N, T>, T> 
		extends	BinaryTrie.Nöde<N, T> {
		int prefix;
		@SuppressWarnings("unchecked")
		public boolean equals(Object u) {
			return (u instanceof Nöde<?, ?>)
					&& this.prefix == ((Nöde<N,T>)u).prefix;
		}
		public int hashCode() {
			return prefix;
		}
	}

	@SuppressWarnings("unchecked")
	public XFastTrie(Node sampleNode, Integerizer<T> it)  {
		super(sampleNode, it);
		t = (USet<Node>[])new USet[w+1];
		Node nil = (Node)new XFastTrie.Nöde<Node,T>();
		nil.prefix = Integer.MIN_VALUE;
		for (int i = 0; i <= w; i++) {
			t[i] = new LinearHashTable<Node>(nil);
		}
		t[0].add(r);
	}

	@SuppressWarnings("unchecked")
	public XFastTrie(Integerizer<T> it)  {
		this((Node)new XFastTrie.Nöde<Node,T>(), it);
	}
	
	public boolean add(T x) {
		if (super.add(x)) {
			int i, c = 0, ix = it.intValue(x);
			Node u = r.child[(ix>>>w-1) & 1];
			for (i = 1; i <= w; i++) {
				u.prefix = ix >>> w-i;
				t[i].add(u);
				c = (ix >>> w-i-1) & 1;
				u = u.child[c];
			}
			return true;
		}
		return false;
	}
	
	public boolean remove(T x) {
		// 1 - find leaf, u, containing x
		int i = 0, c, ix = it.intValue(x);
		Node u = r;
		for (i = 0; i < w; i++) {
			c = (ix >>> w-i-1) & 1;
			if (u.child[c] == null) return false;
			u = u.child[c];
		}
		// 2 - remove u from linked list
		Node pred = u.child[prev];   // predecessor
		Node succ = u.child[next];   // successor
		pred.child[next] = succ;
		succ.child[prev] = pred;
		u.child[next] = u.child[prev] = null;
		Node w = u;
		// 3 - delete nodes on path to u
		while (w != r && w.child[left] == null && w.child[right] == null) {
			if (w == w.parent.child[left])
				w.parent.child[left] = null;
			else // u == u.parent.child[right] 
				w.parent.child[right] = null;
			t[i--].remove(w);
			w = w.parent;
		}
		// 4 - update jump pointers
		w.jump = (w.child[left] == null) ? succ : pred;
		w = w.parent;
		while (w != null) {
			if (w.jump == u)
				w.jump = (w.child[left] == null) ? succ : pred;
			w = w.parent;
		}
		n--;
		return true;
	}
	
	protected Node findNode(int ix) {
		// find lowest node that is an ancestor of ix
		int l = 0, h = w+1;
		Node v, u = r, q = newNode();
		while (h-l > 1) {
			int i = (l+h)/2;
			q.prefix = ix >>> w-i;
			if ((v = t[i].find(q)) == null) {
				h = i;
			} else {
				u = v;
				l = i;
			}
		}
		if (l == w) return u;
		Node pred = (((ix >>> w-l-1) & 1) == 1) ? u.jump : u.jump.child[0];
		return (pred.child[next] == dummy) ? null : pred.child[next];
	}
	
	public void clear() {
		super.clear();
		for (USet<Node> s : t) 
			s.clear();
	}
	
	public T find(T x) {
		int l = 0, h = w+1, ix = it.intValue(x);
		Node v, u = r, q = newNode();
		while (h-l > 1) {
			int i = (l+h)/2;
			q.prefix = ix >>> w-i;
			if ((v = t[i].find(q)) == null) {
				h = i;
			} else {
				u = v;
				l = i;
			}
		}
		if (l == w) return u.x;
		Node pred = (((ix >>> w-l-1) & 1) == 1) 
			     ? u.jump : u.jump.child[0];
		return (pred.child[next] == dummy) 
		             ? null : pred.child[next].x;
	}
	
	public static void main(String[] args) {
		class N extends XFastTrie.Nöde<N, Integer> {};
		class I implements Integerizer<Integer> { 
			public int intValue(Integer i) { return i; }
		}
		XFastTrie<N, Integer> t = new XFastTrie<N, Integer>(new I()); 
		BinaryTrie.easyTests(t, 20);
	}
}

File: python/ods/xfasttrie.py
"""An implementation of Willard's X-Fast tries

This structure is able to store w-bit integers with O(log w) time searches
and O(w) time addition/removal

D. E. Willard. Log-logarithmic worst-case range queries are possible in 
  space Theta(n). Information Processing Letters, 17, 81-84. 1984.
"""
from base import BaseSet
from binarytrie import BinaryTrie
from linearhashtable import LinearHashTable
from utils import new_array, w


class XFastTrie(BinaryTrie):
    class Node(BinaryTrie.Node):
        def __init__(self):
            super(XFastTrie.Node, self).__init__()
            self.prefix = 0
            
        def __eq__(self, other):
            return self.prefix == other.prefix
        
        def __hash__(self):
            return hash(self.prefix)
        
    def _new_node(self):
        return XFastTrie.Node()
        
    def __init__(self):
        super(XFastTrie, self).__init__()
        self.nil = self._new_node()
        self.t = new_array(w+1)
        for i in range(w+1):
            self.t[i] = LinearHashTable()
        self.t[0].add(self.r)
    
    def add(self, x):
        if super(XFastTrie, self).add(x):
            ix = int(x)
            u = self.r.child[(ix >> w-1) & 1]
            for i in range(1, w+1):
                u.prefix = ix >> w-i
                self.t[i].add(u)
                if i < w:
                    c = (ix >> w-i-1) & 1
                u = u.child[c]
            return True
        return False
    
    def find(self, x):
        ix = int(x)
        l, h = 0, w+1
        u = self.r
        q = self._new_node()
        while h-l > 1:
            i = (l+h)/2
            q.prefix = ix >> w-i
            v = self.t[i].find(q)
            if v is None:
                h = i
            else:
                u = v
                l = i
        if l == w: return u.x
        c = ix >> (w-l-1) & 1
        pred = [u.jump.prev, u.jump][c]
        if pred.next is None: return None
        return pred.next.x

    # TODO: Too much duplication with find(x)
    def find_node(self, x):
        ix = int(x)
        l, h = 0, w+1
        u = self.r
        q = self._new_node()
        while h-l > 1:
            i = (l+h)/2
            q.prefix = ix >> w-i
            v = self.t[i].find(q)
            if v is None:
                h = i
            else:
                u = v
                l = i
        if l == w: return u
        c = ix >> (w-l-1) & 1
        pred = [u.jump.prev, u.jump][c]
        if pred.next is None: return None
        return pred.next
    
    def remove(self, x):
        # 1 - fine leaf, u, containing x
        ix = int(x)
        i = 0
        u = self.r
        while i < w:
            c = ix >> (w-i-1) & 1
            u = u.child[c]
            if u is None: return False
            i += 1
        
        # 2 - remove u from linked list
        pred = u.prev
        succ = u.next
        pred.next = succ
        succ.prev = pred
        u.next = u.prev = None
        v = u
        
        # 3 - delete nodes on path to u
        while v is not self.r and v.left is None and v.right is None:
            if v is v.parent.left:
                v.parent.left = None
            else:
                v.parent.right = None
            self.t[i].remove(v)
            i -= 1
            v = v.parent
        
        # 4 - update jump pointers
        v.jump = [pred, succ][v.left is None]
        v = v.parent
        while v is not None:
            if v.jump is u:
                v.jump = [pred, succ][v.left is None]
            v = v.parent
        self.n -= 1
        return True
            
                
        
      